### PR TITLE
User defined VxLAN UDP source port calculation

### PR DESCRIPTION
**Pull request by marian-pritsak**

### CODE REVIEW

File: inc/saiswitch.h [hunk 1]
```diff
@@ -490,6 +495,32 @@ typedef enum _sai_switch_tunnel_attr_t
      */
     SAI_SWITCH_TUNNEL_ATTR_VXLAN_UDP_SPORT,
 
+    /**
+     * @brief Tunnel UDP source port base for a calculated hash range
+     *
+     * Value of the fixed portion of the UDP source port's MSB
```
Comment (itaibaz) [inline inc/saiswitch.h:501]: perhaps would be more clear to write most significant bits in plural as MSB is most significant bit
same for LSB below

File: inc/saiswitch.h [hunk 2]
```diff
@@ -490,6 +495,32 @@ typedef enum _sai_switch_tunnel_attr_t
...
+     * @validonly SAI_SWITCH_TUNNEL_ATTR_TUNNEL_TYPE == SAI_TUNNEL_TYPE_VXLAN and SAI_SWITCH_TUNNEL_ATTR_TUNNEL_VXLAN_UDP_SPORT_MODE == SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_USER_DEFINED_HASH
+     */
+    SAI_SWITCH_TUNNEL_ATTR_VXLAN_UDP_SPORT_BASE,
+
+    /**
+     * @brief Tunnel UDP source port hash bit length
+     *
+     * Number of LSB reserved for the calculated hash value
+     *
+     * @type sai_uint16_t
```
Comment (itaibaz) [inline inc/saiswitch.h:516]: probably can be uint8_t instead of uint16_t, as we need to describe a value between 0 to 16

**Review by itaibaz — COMMENTED** [review]
should we also add per tunnel attrib ?
as original attributes are both per {switch, tunnel type} and {tunnel}
otherwise, we may need to specify this mode is not per tunnel

**Comment by JaiOCP** [comment]
Have you looked at PR#1173. It added following attributes
/**
 * @brief Defines VXLAN tunnel UDP source port mode
 */
typedef enum _sai_tunnel_vxlan_udp_sport_mode_t
{
    /**
     * @brief User define value
     */
    SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_USER_DEFINED,

    /**
     * @brief RFC6335 Computed hash value in range 49152-65535
     */
    SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_EPHEMERAL,
} sai_tunnel_vxlan_udp_sport_mode_t;

**Comment by marian-pritsak** [comment]
> should we also add per tunnel attrib ?
> as original attributes are both per {switch, tunnel type} and {tunnel}
> otherwise, we may need to specify this mode is not per tunnel

It doesn't make sense at the moment to have it per tunnel. If anyone will come up with the use case, it can be added.

**Comment by marian-pritsak** [comment]
> Have you looked at PR#1173. It added following attributes
> /**
> 
> * @brief Defines VXLAN tunnel UDP source port mode
>   _/
>   typedef enum _sai_tunnel_vxlan_udp_sport_mode_t
>   {
>   /_*
>   
>   * @brief User define value
>     */
>     SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_USER_DEFINED,
>   
>   /**
>   
>   * @brief RFC6335 Computed hash value in range 49152-65535
>     */
>     SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_EPHEMERAL,
>     } sai_tunnel_vxlan_udp_sport_mode_t;

"Ephemeral" is not flexible enough

**Comment by marian-pritsak** [comment]
> > should we also add per tunnel attrib ?
> > as original attributes are both per {switch, tunnel type} and {tunnel}
> > otherwise, we may need to specify this mode is not per tunnel
> 
> It doesn't make sense at the moment to have it per tunnel. If anyone will come up with the use case, it can be added.

I misunderstood a comment - yes, it should be in both since the enum is shared

File: inc/saiswitch.h [hunk 3]
```diff
@@ -354,6 +354,11 @@ typedef enum _sai_tunnel_vxlan_udp_sport_mode_t
      * @brief RFC6335 Computed hash value in range 49152-65535
      */
     SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_EPHEMERAL,
+
+    /**
+     * @brief User defined bit range of computed hash
+     */
+    SAI_TUNNEL_VXLAN_UDP_SPORT_MODE_USER_DEFINED_HASH,
```
Comment (lguohan) [inline inc/saiswitch.h:361]: can be merged into user_defined mode

Comment (marian-pritsak) [reply inc/saiswitch.h]: Done

**Comment by rlhui** [comment]
@JaiOCP  - would you like to review/approve?

**Review by lguohan — APPROVED** [review]

**Review by kcudnik — APPROVED** [review]

**Comment by JaiOCP** [comment]
> @JaiOCP - would you like to review/approve
Changes looks good

**Review by JaiOCP — APPROVED** [review]

**Pull request merged**